platform/mbed_reset_reason.h

```diff
@@ -0,0 +1,56 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_RESET_REASON_H
+#define MBED_RESET_REASON_H
+
+#if DEVICE_RESET_REASON
+
+#include "reset_reason_api.h"
+
+namespace mbed {
+/** \addtogroup platform */
+/** @{*/
+/**
+ * \defgroup reset_reason reset reason functions
+ * @{
+ */
+
+/** Get the platform-independent reason code for the last system reset.
+ *
+ *  Example:
+ *  @code
+ *  const reset_reason_t reason = mbed_reset_reason_get();
+ *
+ *  if (reason == RESET_REASON_WATCHDOG) {
+ *      std::cout << "Watchdog reset" << std::endl;
+ *      rollback();
+ *  }
+ *  @endcode
+ */
+reset_reason_t mbed_reset_reason_get(void);
```

Comment:
A more verbose description is needed regarding the side effects.

> Not wanting to support platform specific things in the HAL is reasonable. That was the intention of the PLATFORM return value. The application can use the platform specific SDK to get the platform specific reset reason after receiving it.

> But, the state of the registers should be tracked/stored somewhere to allow multiple reads. e.g if get clears the registers but returns RESET_REASON_PLATFORM there is no way to read the registers again to determine the reason without storing the register state.

Suppose I got `RESET_REASON_PLATFORM`, `RESET_REASON_ACCESS_ERROR` or `RESET_REASON_BOOT_ERROR` from `mbed_reset_reason_get()`. As @scartmell-arm wrote there is no way to get more specific info since registers got wiped out already. This excludes the use of `mbed_reset_reason_get()` when platform specific SDK is needed.

Reply:
Yeah, that does seem to be an issue. I think that it would be cleared if the registers are cleared when reading, but it does make it problematic if the reason can't be specified on the first read. Perhaps a criterion that says it won't clear if those specific generic reasons are returned, though that does make it more convoluted.

Reply:
I'm not following - how does one arrange for it not to be cleared in the platform-specific SDK, if reading the registers clears them? (Good catch, @fkjagodzinski )

Reply:
Can we clear the registers every time and store it ourselves in the driver?

Reply:
That does have some nice benefits. Only, what if more than one hardware register is involved? I've never seen that, but I haven't seen everything.

Reply:
True. OK, I think we are overthinking this. How about we just do 'you have to map your reset reason to our enum period' if there's a generic usecase we can add it to our list and we will get rid of the platform reason.

Reply:
> That does have some nice benefits. Only, what if more than one hardware register is involved? I've never seen that, but I haven't seen everything. Anybody else seen multi-register reset reasons?

Not that I've seen. K64F reset reasons are stored in two adjacent 8 bit registers, but that could just be stored as one 16 bit one.

> How about we just do 'you have to map your reset reason to our enum period' if there's a generic usecase we can add it to our list and we will get rid of the platform reason.

Would be the simplest solution.

Reply:
That would be simplest, but it screws anyone with non-generic reasons they are interested in. How about halfway - read, use and store the single most relevant register (likely the only one, apparently), and make that available, but encourage mapping to the generic reasons for more portable code? That should satisfy almost everybody with the generics, and the few others have an escape route.

Reply:
if we use uint32_t for storing the status we have plenty of space and the hal implementation can do the right thing.